Editing Toolkit: Add feature flags to the help center

jeyip opened the pull request:
#### Context

The help center is being displayed to the general public. Next Steps tutorials frontend development should be hidden from users until it's ready to be released.

paYE8P-1Hx-p2

#### Proposed Changes

* Adds a backend filter that can be toggled as a feature flag for the next steps tutorial UI
* Adds frontend context to retrieve the flag from an enqueued help center script

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Load this PR
* Sandbox a site
* Run `yarn dev --sync` in `apps/editing-toolkit`
* Import the `useFeatureFlags` hook from `help-center/src/contexts/FeatureFlagContext.tsx` into any functional component
* console log the result of `useFeatureFlags()` in the functional component
* Navigate to the site editor in the sandboxed site
* Navigate to the help center in the sandboxed site ( question mark icon at the top right hand corner of the screen )
* Open the developer console and note the feature flag boolean being logged out
* Update the value of `apps/editing-toolkit/editing-toolkit-plugin/help-center/class-help-center.php:48` to `true`
* Verify that the feature flag boolean has changed values

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/65130

jeyip committed:
Add feature flag filter to help center

Addison-Stavlo commented on packages/help-center/src/contexts/FeatureFlagContext.tsx:
```diff
@@ -0,0 +1,20 @@
+} );
+
+export const FeatureFlagProvider: React.FC< FeatureFlags > = function ( { children } ) {
+	// The helpCenterFeatureFlags value is added as a global through the help-center-script
+	return (
+		<FeatureFlagContext.Provider value={ helpCenterFeatureFlags }>
```
This is non-blocking and more for curious discussion.

Since `helpCenterFeatureFlags` is a global, what is the point of setting up a context provider and checking the result of the `useFeatureFlags` hook.. when we could just check the `helpCenterFeatureFlags` global instead anyways?  Honestly, I really like being able to use `useFeatureFlags` more than a global... it just feels a lot more clean. Curious about your thoughts though?

jeyip replied:
> Since helpCenterFeatureFlags is a global, what is the point of setting up a context provider and checking the result of the useFeatureFlags hook

Good question! The context + provider allows us to consolidates direct references to the global to a single spot. We could've done this in many ways, but since feature flags are typically used at different levels of the component tree, context seemed like an idiomatic way to make this information available without all the prop drilling.

Since feature flags will probably be reused throughout the codebase at some point, the useFeatureFlags hook is partly for dev ergonomics so that we don't have to repeatedly import useContext and FeatureFlagContext, and partly so that we can add error handling ~( which I haven't done yet )~

```js
function useFeatureFlags() {
    const context = React.useContext(FeatureFlagContext)
   
    // Trying to call hook outside of component that renders FeatureFlagProvider
    if (context === undefined) {
      throw new Error('useFeatureFlags must be used within a FeatureFlagProvider')
    }
    return context
}
```

I learned about it here https://kentcdodds.com/blog/how-to-use-react-context-effectively

Addison-Stavlo reviewed (commented):
This tests well from my end! 🥳 

Are there any other things you plan to resolve here (curious since its still in 'draft' mode)?

jeyip commented:
> Are there any other things you plan to resolve here (curious since its still in 'draft' mode)?

Yeah two things 👍 

- I saw some failing CI specs before I threw this up yesterday, but had to leave in a hurry. (Might be unrelated)
- Addressing the typescript errors I was seeing in the client side files

jeyip committed:
Add error handling to useFeatureFlags context

jeyip committed:
Add explicit react import to avoid errors

jeyip committed:
Declare feature flags global variable in typescript

jeyip committed:
Fix FeatureFlagProvider type declaration

jeyip commented on packages/help-center/src/components/help-center-container.tsx:
```diff
@@ -54,30 +55,32 @@ const HelpCenterContainer: React.FC< Container > = ( { handleClose, isLoading }
-					) }
-				</Card>
-			</OptionalDraggable>
+			<FeatureFlagProvider>
```
The only meaningful change in this file is the addition of `FeatureFlagProvider`

agrullon95 commented on packages/help-center/src/contexts/FeatureFlagContext.tsx:
```diff
@@ -0,0 +1,31 @@
+import * as React from 'react';
```
Thanks for adding `react` import as we discussed 👍

agrullon95 commented on packages/help-center/src/contexts/FeatureFlagContext.tsx:
```diff
@@ -0,0 +1,31 @@
+	const featureFlags = useContext( FeatureFlagContext );
+
+	// Trying to call hook outside of component that renders FeatureFlagProvider
+	if ( featureFlags === undefined ) {
```
I have imported the `useFeatureFlags` hook it to the products purchase page (`/product-purchase-features-list/customize-theme`outside of the FeatureFlagProvider) and I'm still able to see the flag value. Can you re-test this and let me know if you're able to replicate it?

jeyip replied:
Good catch! I tested things out again the problem should be fixed in [514e133](https://github.com/Automattic/wp-calypso/pull/65142/commits/514e133047d5b47cbceb1093419800bafe4ce691)

agrullon95 replied:
Tested and it's working now 👍

jeyip committed:
Throw error when useFeatureFlags hook is rendered improperly

Ensure that useFeatureFlags hook is only used within components rendered within the FeatureFlagProvider

escapemanuele reviewed (approved):
Thanks for providing feature flags to the help center 👍

jeyip committed:
Fix php linting error
